### PR TITLE
[WIP] [CARBONDATA-1592] Adding event listener interface to Carbondata

### PR DESCRIPTION
**Pull request by manishgupta88**
Adding event listener interface to Carbondata

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/462/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 1.6, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/337/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1090/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 1.6, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/340/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/465/

File: core/src/main/java/org/apache/carbondata/core/datamap/TableDataMap.java [hunk 1]
```diff
@@ -160,4 +156,12 @@ public String getDataMapName() {
   public DataMapFactory getDataMapFactory() {
     return dataMapFactory;
   }
+
+  @Override public void onEvent(Event event) {
+    dataMapFactory.fireEvent(event);
+  }
+
+  @Override public void onAbort(Event event) {
```
Comment (gvramana) [inline core/src/main/java/org/apache/carbondata/core/datamap/TableDataMap.java]: this method is not required in Event Interface as Abort is a separate event.

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1093/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/468/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 1.6, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/343/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1096/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 1.6, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/350/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/475/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1103/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/476/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1104/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/478/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/353/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1106/

File: integration/spark2/src/main/scala/org/apache/carbondata/events/Events.scala [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+ */
+trait TableEvent extends DatabaseEvent {
+  val carbonTableIdentifier: CarbonTableIdentifier
+  override lazy val databaseName: String = carbonTableIdentifier.getDatabaseName
+}
+
+/**
+ * event for load operations
+ */
+trait LoadEvent extends TableEvent {
```
Comment (gvramana) [inline integration/spark2/src/main/scala/org/apache/carbondata/events/Events.scala]: Move this trait to LoadEvents class

File: integration/spark2/src/main/scala/org/apache/carbondata/events/EventListenerExample.scala [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+  val warehouse = s"$rootPath/examples/spark2/target/warehouse"
+  val sparkSession = SparkSession
+    .builder()
+    .master("local")
+    .appName("CarbonSessionExample")
+    .config("spark.sql.warehouse.dir", warehouse)
+    .config("spark.driver.host", "localhost")
+    .getOrCreateCarbonSession(storeLocation)
+  val carbonTableIdentifier = new CarbonTableIdentifier("db1", "tbl1", "tbl1-id")
+  val carbonLoadModel = new CarbonLoadModel
```
Comment (gvramana) [inline integration/spark2/src/main/scala/org/apache/carbondata/events/EventListenerExample.scala]: Example to be used how to subscribe to data load event

**Comment by jackylk** [comment]
Can you explain more on the intention of this PR?

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1175/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/544/

**Comment by manishgupta88** [comment]
@jackylk ..This PR will provide an event listener interface using which you can register the events based on the need of your operations. For example  before dropping a table you need to record the information from that table that can be used for clean up of resources/caches used by that table on executor or driver side.  In this case you can register an event PreDropEventListener which will perform the required task and add an event which will triggered just before drop.

File: integration/spark2/src/main/scala/org/apache/carbondata/events/Events.scala (hunk 2 above)
Comment (manishgupta88) [reply integration/spark2/src/main/scala/org/apache/carbondata/events/Events.scala]: I think it is better to keep all the possible events in Events.scala class in order to know what all different type of events are possible and keep their concrete implementation in different classes to clearly define their behavior.

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1284/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/650/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/854/

**Comment by manishgupta88** [comment]
Code already merged as part of PR #1473

**Pull request closed without merging**